rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/geo/GeofencingProcessor.java

```diff
@@ -0,0 +1,198 @@
+        Set<GeofenceState> candidatesForOutside = geofenceStates.stream().filter(state -> state.isLeft() && !matchedGeofences.contains(state.getGeofenceId())).collect(Collectors.toSet());
+
+        if (CollectionUtils.isEmpty(candidatesForOutside)) {
+            return emptyList();
+        }
+
+        List<GeofenceState> outsideGeofences = new ArrayList<>();
+
+        for (GeofenceState candidate : candidatesForOutside) {
+            long diff = currentTs - candidate.getLeftTs();
+            boolean isTimeForOutside = diff > durationMs;
+            if (isTimeForOutside) {
+                outsideGeofences.add(candidate);
+            }
+        }
+
+        outsideGeofences.forEach(geofenceStates::remove);
+
+        return outsideGeofences.stream().map(GeofenceState::getGeofenceId).collect(Collectors.toList());
+    }
+
+    public boolean hasEnteredGeofence(List<EntityId> matchedGeofences, Set<GeofenceState> geofenceStates) {
+        return geofenceStates.stream().anyMatch(geofenceState -> geofenceState.isEntered() && matchedGeofences.contains(geofenceState.getGeofenceId()));
+    }
+
+    public boolean hasEnteredNotMatchedGeofence(List<EntityId> matchedGeofences, Set<GeofenceState> geofenceStates) {
+        return geofenceStates.stream().anyMatch(geofenceState -> geofenceState.isEntered() && !matchedGeofences.contains(geofenceState.getGeofenceId()));
+    }
+
+    public boolean hasNotEnteredAnyGeofence(Set<GeofenceState> geofenceStates) {
+        return geofenceStates.stream().allMatch(GeofenceState::isLeft);
+    }
+
+    private void persistState(EntityId entityId, RuleNodeId ruleNodeId, Set<GeofenceState> geofenceStates) {
+        ObjectNode jsonNode = JacksonUtil.newObjectNode();
+
+        jsonNode.set(GEOFENCE_STATES_KEY, JacksonUtil.valueToTree(geofenceStates));
+
+        BaseAttributeKvEntry attributeKvEntry = new BaseAttributeKvEntry(new JsonDataEntry(getGeofenceStateAttributeKey(ruleNodeId), jsonNode.toString()), System.currentTimeMillis());
+        context.getAttributesService().save(context.getTenantId(), entityId, AttributeScope.SERVER_SCOPE, attributeKvEntry);
+    }
+
+    private void deleteGeofenceStateByGeofenceId(EntityId geofenceId, Set<GeofenceState> geofenceStates) {
+        Optional<GeofenceState> geofenceState = geofenceStates.stream().filter(state -> state.getGeofenceId().equals(geofenceId)).findFirst();
+        geofenceState.ifPresent(geofenceStates::remove);
+    }
+
+    public Set<GeofenceState> fetchGeofenceStatesForEntity(EntityId entityId, RuleNodeId ruleNodeId) throws ExecutionException, InterruptedException {
+        Optional<AttributeKvEntry> attributeKvEntry = context.getAttributesService().find(context.getTenantId(), entityId, AttributeScope.SERVER_SCOPE, getGeofenceStateAttributeKey(ruleNodeId)).get();
+        if (attributeKvEntry.isEmpty() || attributeKvEntry.get().getJsonValue().isEmpty()) {
+            return new HashSet<>();
```

Comment:
Maybe it's better to use Collections.emptySet() or null?

Reply:
1. emptySet is not appropriate option here because it results in
```
java.lang.UnsupportedOperationException
	at java.base/java.util.AbstractCollection.add(AbstractCollection.java:253)
	at org.thingsboard.rule.engine.geo.GeofencingProcessor.getEnteredGeofences(GeofencingProcessor.java:99)
	at org.thingsboard.rule.engine.geo.GeofencingProcessor.lambda$process$0(GeofencingProcessor.java:73)
	at com.google.common.util.concurrent.AbstractTransformFuture$TransformFuture.doTransform(AbstractTransformFuture.java:254)
	at com.google.common.util.concurrent.AbstractTransformFuture$TransformFuture.doTransform(AbstractTransformFuture.java:244)
	at com.google.common.util.concurrent.AbstractTransformFuture.run(AbstractTransformFuture.java:125)
	at com.google.common.util.concurrent.DirectExecutor.execute(DirectExecutor.java:31)
	at com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:1298)
	at com.google.common.util.concurrent.AbstractFuture.addListener(AbstractFuture.java:762)
	at com.google.common.util.concurrent.FluentFuture$TrustedFuture.addListener(FluentFuture.java:114)
	at com.google.common.util.concurrent.AbstractTransformFuture.create(AbstractTransformFuture.java:53)
	at com.google.common.util.concurrent.Futures.transform(Futures.java:490)
	at org.thingsboard.rule.engine.geo.GeofencingProcessor.process(GeofencingProcessor.java:60)
	at org.thingsboard.rule.engine.geo.TbGpsMultiGeofencingActionNode.lambda$onMsg$0(TbGpsMultiGeofencingActionNode.java:71)
	at com.google.common.util.concurrent.AbstractTransformFuture$AsyncTransformFuture.doTransform(AbstractTransformFuture.java:225)
	at com.google.common.util.concurrent.AbstractTransformFuture$AsyncTransformFuture.doTransform(AbstractTransformFuture.java:212)
	at com.google.common.util.concurrent.AbstractTransformFuture.run(AbstractTransformFuture.java:125)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1423)
	at java.base/java.util.concurrent.ForkJoinTask.doExec$$$capture(ForkJoinTask.java:387)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1311)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1841)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1806)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)

```

2. Null is not option because it's easier to handle empty collection rather than write many null checks